File: leader.py
def solution(A):
    if len(A) == 1:
        return 0
    
    leader = find_leader(A)
    if leader is None:
        return 0
    
    left = left_leaders(A, leader)
    right = right_leaders(A, leader)
    count = 0
    breakpoint()
    for l, r in zip(left, right):
        if l and r:
            count += 1
    return count


def find_leader(A):
    size, top = 0, None
    for x in A:
        if size == 0:
            top = x
            size += 1
        else:
            if x == top:
                size += 1
            else:
                size -= 1
    count, n = 0, len(A)
    candidate = top
    for x in A:
        if x == candidate:
            count += 1
            if count > (n // 2):
                return candidate
    return None


def left_leaders(A, leader):
    curr, same = 0, []
    A = A[:-1]
    for length, x in enumerate(A, 1):
        if x == leader:
            curr += 1
        same.append(curr > (length // 2))
    return same


def right_leaders(A, leader):
    curr, same = 0, []
    A = A[1:]
    for length, x in enumerate(list(reversed(A)), 1): 
        if x == leader:
            curr += 1
        same.append(curr > (length // 2))
    same.reverse()
    return same


def test():
    assert solution([4, 3, 4, 4, 4, 2]) == 2


if __name__ == '__main__':
    test()

File: leader.c
int solution(int A[], int N) {
    int candidate;
    int count = 0;
    int i;
    // 1. Find most likely candidate for the leader
    for(i = 0; i < N; i++){
        // change candidate when count reaches 0
        if(count == 0) candidate = i;
        // count occurrences of candidate
        if(A[i] == A[candidate]) count++;
        else count--;
    }
    // 2. Verify that candidate occurs more than N/2 times
    count = 0;
    for(i = 0; i < N; i++) if(A[i] == A[candidate]) count++;
    // If there's no leader then that there are also no equi-leaders
    if (count <= N/2) return 0;
    // 3. There exists a leader now count the number of occurences
    // of this leader in all sequences.
    int leader = A[candidate];
    int equiLeaderCount = 0;
    int cntL = 0; // leader count for the left sequence
    int cntR = count; // leader count for the right sequence
    for(int i = 0; i < N-1; i++){
        if(A[i] == leader){
            cntL++;
            cntR--;
        }
        // verify that the dominant value in both sequences
        // is the same
        if(cntL > (i+1)/2 && cntR > (N-i-1)/2) equiLeaderCount++;
    }
    return equiLeaderCount;
}
